protocol-units/settlement/mcr/contracts/src/settlement/MCR.sol

```diff
@@ -149,15 +149,39 @@ contract MCR is Initializable, BaseSettlement, MCRStorage, IMCR {
 
         emit BlockCommitmentSubmitted(blockCommitment.blockId, blockCommitment.commitment, allCurrentEpochStake);
 
+    }
+
+    /// @notice The current leader can attest to a block height, given there is a supermajority of stake on the block
+    function attestBlocks(address attester) public {
+        // check if the address is the current leader
+        if (attester != getCurrentLeader()) revert("NotLeader");
+
         // keep ticking through to find accepted blocks
         // note: this is what allows for batching to be successful
         // we can commit to blocks out to the tolerance point
         // then we can accept them in order
-        // ! however, this does potentially become very costly for whomever submits this last block
-        // ! rewards need to be managed accordingly
+        // ! rewards need to be 
+        // ! - at least proportional to attested blocks to account for consumed gas
+        // ! - reward the leader well to incentivize frequent block attestation (close to comitted block frequency)
+        //     rather than incentivizing the leader to batch attesting blocks
         while (tickOnBlockHeight(lastAcceptedBlockHeight + 1)) {}
     }
 
+    function getCurrentLeader() public view returns (address) {
+        // TODO replace the following with a configurable value
+        // the leader remains the same for leaderterm L1-blocks
+        uint256 leaderTerm = 100;
+
+        uint256 currentL1BlockHeight = block.number;
+        uint256 relevantL1BlockHeight = currentL1BlockHeight - currentL1BlockHeight % leaderTerm ;
+        bytes32 blockHash = blockhash(relevantL1BlockHeight);
+
+        address[] memory attesters = getAttesters();
+        // map the blockhash to the attesters
+        uint256 leaderIndex = uint256(blockHash) % attesters.length;
```

Comment:
I'm sure this is in the MDs that I haven't gotten to you, but doesn't this particular election allow a leader to just mine for a blockhash that would elect themselves. Really, we would want quorum-based leader election if we are going to introduce a leader.

Reply:
It's L1 blockhash not L2. Will change the variable name to clarify

Reply:
(Long term I think this is also not a great assumption but it buys time to look into a better randomness solution )